File: 백준/Gold/1717. 집합의 표현/집합의 표현.py
'''
union-find (disjoint set)
입력에 따라 그때그때 실행하면서 출력해주면 된다.
'''
import sys
input = sys.stdin.readline


def find(p):
    if parent[p] != p:
        parent[p] = find(parent[p])
    return parent[p]


def union(p, q):
    pp = find(p)
    pq = find(q)
    if pp == pq:
        return
    if rank[pp] <= rank[pq]:
        parent[pp] = pq
        if rank[pp] == rank[pq]:
            rank[pq] += 1
    elif rank[pq] < rank[pp]:
        parent[pq] = pp


n, m = map(int, input().split())
parent = [i for i in range(n+1)]
rank = [0] * (n+1)
depth = 0
for _ in range(m):
    command, p, q = map(int, input().split())
    if command:
        if find(p) == find(q):
            print("YES")
        else:
            print("NO")
    else:
        union(p, q)

File: 백준/Gold/1717. 집합의 표현/집합의 표현.cc
#include <iostream>
using namespace std;

int N, M;
int parent[1000001];
int level[1000001];


int Find(int p) {
    if (parent[p] != p) parent[p] = Find(parent[p]);
    return parent[p];
}

void Union(int p, int q) {
    int pp = Find(p);
    int pq = Find(q);
    if (level[pp] <= level[pq]) {
        parent[pq] = pp;
        level[pq]++;
    }
    else {
        parent[pq] = pp;
        level[pp]++;
    }
}

int main(void) {
    ios_base :: sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
    cin >> N >> M;
    for (int i = 0; i <= N; i++) {
        parent[i] = i;
        level[i] = 1;
    }
    for (int i = 0; i < M; i++) {
        int command, p, q;
        cin >> command >> p >> q;
        if (command) {
            if (Find(p) == Find(q)) cout << "YES\n";
            else cout << "NO\n";
        } else {
            Union(p, q);
        }
    }
}